### PR TITLE
Making `test_winreg` passes.

### LINKED ISSUE #410: test_winreg is blocking
The following code isn't terminating:
```Python
import threading

def test_thread():
    class Thread(threading.Thread):
        def run(self):
            print("run")

    thread = Thread()
    thread.start()
    print("join")
    thread.join()
    print("done")
    
test_thread()
```

There are some TODOs in `_thread.cs` so that might be the issue.

### PR DESCRIPTION
**Pull request by in-code-i-trust**
Hi, 
I wrote a patch to resolve #410.  

The code shown in #410 causes a deadlock because `_set_sentinel()` is not implemented as `threading.py` expects; an lock object returned by `_set_sentinel()` should be released on its thread exit.  
  
I also implemented a lacking feature, timeout for `acquire()` as well. 
According to [the documentation](https://docs.python.org/3/library/threading.html#threading.Lock.acquire), timeout is only supported on blocking acquire.

### CODE REVIEW

File: Src/IronPython.Modules/_thread.cs [hunk 1]
```diff
@@ -23,6 +24,7 @@ public static class PythonThread {
 
         private static readonly object _stackSizeKey = new object();
         private static object _threadCountKey = new object();
+        private static ConcurrentDictionary<int, Lazy<@lock>> _sentinelTable = new ConcurrentDictionary<int, Lazy<@lock>>();
```
Comment (slozier) [inline Src/IronPython.Modules/_thread.cs]: Might be a good candidate for a `ThreadStatic`?
```C#
[ThreadStatic]
private static @lock _sentinel;
```

**Comment by in-code-i-trust** [comment]
Okay..., this PR is completely messed up. I’m sorry. I will fix it. 
Thank you for rerunning tests and the review comments.

**Comment by in-code-i-trust** [comment]
@slozier 
I've fixed the problems and enabled `test_winreg`.  As you mentioned, the test code at #410 is now in `test_regressions`.  
  
So here are what i've done in this PR.  
- Threads can now have multiple sentinel locks. 
  - This makes `test-all` passes.
- Implemented some lacking features in `winreg.cs`.
  - This makes `test_winreg` passes, so it's been enabled.
- Ported the codes shown in #410 to `test_regressions.py`.

File: Src/IronPython.Modules/_thread.cs (hunk 1 above)
Comment (in-code-i-trust) [reply Src/IronPython.Modules/_thread.cs]: Never knew that. Replaced with it in commit 972a7bd.

**Pull request merged**